lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -367,6 +428,322 @@ HdVP2Mesh::HdVP2Mesh(HdVP2RenderDelegate* delegate, const SdfPath& id, const Sdf
+        // Iterate through all reprdescs for the current repr to figure out if any
+        // of them requires smooth normals or flat normals. If either (or both)
+        // are required, we will calculate them once and clean the bits.
+        bool requireSmoothNormals = false;
+        bool requireFlatNormals = false;
+        int  drawItemIndex = 0;
+        for (size_t descIdx = 0; descIdx < reprDescs.size(); ++descIdx) {
+            const HdMeshReprDesc& desc = reprDescs[descIdx];
+            if (desc.geomStyle == HdMeshGeomStyleHull) {
+                if (desc.flatShadingEnabled) {
+                    requireFlatNormals = true;
+                } else {
+                    requireSmoothNormals = true;
+                }
+            }
+        }
+
+        // If there are authored normals, prepare buffer only when it is dirty.
+        // otherwise, compute smooth normals from points and adjacency and we
+        // have a custom dirty bit to determine whether update is needed.
+        if (requireSmoothNormals && (rprimDirtyBits & DirtySmoothNormals)) {
+#ifdef HDVP2_ENABLE_GPU_COMPUTE
+            if (_gpuNormalsEnabled) {
+                if (!_meshSharedData->_viewportCompute) {
+                    _CreateViewportCompute(*drawItem);
+#ifdef HDVP2_ENABLE_GPU_OSD
+                    _CreateOSDTables();
+#endif
+                }
+                _meshSharedData->_viewportCompute->setNormalVertexBufferGPUDirty();
+                updateNormals
+                    = false; // mark rprimDirtyBits clean? Add something to the update code?
+            } else
+#endif
+            {
+                // note: normals gets dirty when points are marked as dirty,
+                // at change tracker.
+                Hd_VertexAdjacencySharedPtr adjacency(new Hd_VertexAdjacency());
+                HdBufferSourceSharedPtr     adjacencyComputation
+                    = adjacency->GetSharedAdjacencyBuilderComputation(&_meshSharedData->_topology);
+                adjacencyComputation->Resolve();
+
+                // Only the points referenced by the topology are used to compute
+                // smooth normals.
+                VtValue normals(Hd_SmoothNormals::ComputeSmoothNormals(
+                    adjacency.get(),
+                    _points(_meshSharedData->_primvarInfo).size(),
+                    _points(_meshSharedData->_primvarInfo).cdata()));
+
+                if (!normalsInfo) {
+                    _meshSharedData->_primvarInfo[HdTokens->normals]
+                        = std::make_unique<PrimvarInfo>(
+                            PrimvarSource(normals, HdInterpolationVertex, true), nullptr);
+                } else {
+                    normalsInfo->_source.data = normals;
+                    normalsInfo->_source.interpolation = HdInterpolationVertex;
+                }
+            }
+        }
+
+        if (requireFlatNormals && (rprimDirtyBits & DirtyFlatNormals)) {
```

Comment:
In the original code there are these checks for flat normals, but nothing can actually produce the flat normals.